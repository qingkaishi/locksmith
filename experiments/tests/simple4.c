/* Initializes a lock in a function called from main().  Race on 
   non_protected. */
#include <pthread.h>

pthread_mutex_t lock;
int shared;
int non_protected;

void *run(void *arg) {
  pthread_mutex_lock(&lock);
  shared++;
	pthread_mutex_unlock(&lock);
	non_protected++;
//	*(int*) arg = 1;
  return NULL;
}
void doit(int *j) {
  pthread_t t1, t2;
  pthread_mutex_init(&lock, NULL);
  pthread_create(&t1, NULL, run, j);
  pthread_create(&t1, NULL, run, j);
}

int main() {
	int i = 0;
	doit(&i);
  return 1;
}
