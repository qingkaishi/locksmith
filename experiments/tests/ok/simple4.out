************************* STARTING *************************

shared: &non_protected:simple4.c:7
        &shared:simple4.c:6
&shared:simple4.c:6 is protected by:
  concrete lock::-1 -> doit simple4.c:26
  concrete lock:simple4.c:19
  lock:simple4.c:5

Warning: Possible data race: &non_protected:simple4.c:7 is not protected!
 references:
  dereference of &non_protected:simple4.c:7 at simple4.c:13
    &non_protected:simple4.c:7
  locks acquired:
    <empty>
  in: main at simple4.c:24  -> simple4.c:26


*************************** DONE ***************************
